pkg/cachemanager/cachemanager.go

```diff
@@ -201,12 +204,17 @@ func (c *CacheManager) RemoveSource(ctx context.Context, sourceKey aggregator.Ke
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
-	if err := c.gvksToSync.Remove(sourceKey); err != nil {
-		return fmt.Errorf("internal error removing source: %w", err)
-	}
+	currentGVKsForKey := c.gvksToSync.List(sourceKey)
+	gvksNotShared := c.gvksToSync.ListNotShared(sourceKey)
 
+	c.gvksToSync.Remove(sourceKey)
 	err := c.replaceWatchSet(ctx)
-	if general, _ := interpretErr(err, []schema.GroupVersionKind{}); general {
+	if general, failedGVKs := interpretErr(err, gvksNotShared); general || len(failedGVKs) > 0 {
```

Comment:
A couple refinements to the above algorithm:

* We should update the aggregate error object that watch manager returns to specify if any of the errors are associated with removing watches. If none of them are, then we don't need to start the above "retry" goroutine

* To avoid stale data, the retry goroutine will need to mark a cache wipe as being needed.

Reply:
> Looking at the code, I don't think error filtering is the right approach here. 

ok so this goroutine spin-up would only happen in the case of " if any of the errors are associated with removing watches". If so, the routine could be spun up from UpsertSource too since upserting a source can lead to a watch removal too.

> If there is an error on replaceWatchSet, spin up a goroutine ... There should only be one instance of this goroutine at a time.

Given the caveat w UpsertSource leading to watch removal too potentially, why not just have a running go routine that performs some such reconciliation (ie call `replaceWatchSet`)? 

We will keep the direct call to `replaceWatchSet` in `UpsertSource` so watches can get established as soon as possible and enhance the err interpretation to not error out on "watches being removed" for that source.

Then in the `manageCache()` background go routine we can add a `replaceWatchSet`:

```
		case <-c.backgroundManagementTicker.C:
			func() {
				c.mu.Lock()
				defer c.mu.Unlock()
                                 
                                 // TODO future replaceWatchSet call
				c.wipeCacheIfNeeded(ctx)
```

What do you think about this proposal?

Reply:
> ok so this goroutine spin-up would only happen in the case of " if any of the errors are associated with removing watches". If so, the routine could be spun up from UpsertSource too since upserting a source can lead to a watch removal too.

It could, but it'd be unnecessary. We have a mechanism for retrying failed upserts (the reconciler/controller). Since it isn't calling "Remove", we can assume that the error, if it is germane to the sync source being updated, will result in a retry.

Trying to use the same approach for deletes wouldn't work because there is no inbound state, which makes filter-relevant-errors impossible. Instead, we need another way to avoid excess retries. Another option would have been to have a placeholder "delete" object that the controller would retry with for any failed delete event, but that's a slightly higher/weirder coupling than my suggestion.

> What do you think about this proposal?

Worth a try, but it needs a solution for keeping track of when to wipe stale data (wipe needs to be triggered any time a dangling "stop-watch" gets rectified.

Iterable cached data will simplify this flow significantly.

Reply:
TBH setting a flag of "has dangling removals"  would catch this, it could also gate whether replaceWatches is called

Reply:
Lmk what you think of how I tried to use `danglingWatches` flag

Reply:
> It could, but it'd be unnecessary. We have a mechanism for retrying failed upserts (the reconciler/controller). Since it isn't calling "Remove", we can assume that the error, if it is germane to the sync source being updated, will result in a retry.

I take it back, we do want to do retries on removal failure for update -- subsequent reconciles would filter out "irrelevant" GVKs, which would short-circuit the reconcile loop.

> Given the caveat w UpsertSource leading to watch removal too potentially, why not just have a running go routine that performs some such reconciliation (ie call replaceWatchSet)?

^^^ this comment was apt.

Reply:
summarizing this comment thread and some offline conversations for a tl;dr --

---

CacheManager *should own the retrying of watches*, whether to add or remove a watch. In particular, we will retry watches failing to be removed from the `manageCache()` loop. For better, faster latency, we will also err out from UpsertSource if there are any GVKs for which we cannot add watches. In that case,  the controller to retry / re-queue the request. This will also put us in a better place to report status on the SyncSet.

However, we are not going to rely on the controller to retry errors due to watches failing to be removed. This mostly due to A) separation of concerns and B) creating that functionality would require us needing to "restore" the contents of the aggregator for any gvks for which we fail to either add or remove a watch. This adds unneeded complexity that can lead to drift and complicates the control flow needlessly.